Game - first step

Sovianum opened the pull request.

Commit:
Started writing game mechanics

Commit:
Added first attempt to create game process

Commit:
Set semicolons

Commit:
1. Changed project structure.
2. Added regexp length check.

Commit:
./core/lib => ./core/_lib

Commit:
removed libs from tracking

Commit:
Removed node_modules from tracking

Commit:
Removed core/game_lib.js from tracking (temporary)

Commit:
Repaired test

Commit:
Update url

Commit:
Update url and interface (login)

Commit:
Add folder to jshintignore

Commit:
Update carousel js

Antonboom commented on core/server_lib.js:
```diff
@@ -148,19 +148,20 @@ function getStaticServer (router) {
-            (url) => { return request.url.match(new RegExp(url))}
-        );
+        let regexpURL = Object.keys(router.getRegexpURLDict).filter(
+            (url) => { return request.url.match(new RegExp(url)); }
```
match внутри сам преобразовывает входной параметр к RegExp
для справки, не более

Sovianum replied:
Не знал.

Antonboom commented on html/about.html:
```diff
@@ -0,0 +1,51 @@
+
+  <!-- Bootstrap core CSS -->
+  <link href="/lib/bootstrap.min.css" rel="stylesheet">
+  <link href="/lib/bootstrap.css" rel="stylesheet">
```
bootstrap.css перекрывает bootstrap.min.css - может, стоит выбрать что-то одно?

Meganster replied:
Посмотрю и выпилю лишнее

Antonboom commented on html/game_logic_test.html:
```diff
@@ -0,0 +1,20 @@
+    <script src="/core/lib/math.js" type="text/javascript"></script>
+    <script src="/core/game_lib.js"></script>
+    <script>
+        let loop = new GameLoop(document.getElementById("game"), 4, 100, 10);
```
Вынести в game.js, внутри которого инклудить библиотеки, что выше и по отрисовке страницы запускать игру

Sovianum replied:
Извини, а как это сделать?
При загрузке страницы создавать в head новый <script> с нужным путем?
А то require же не работает.

Antonboom replied:
У нас же до сих пор нет webpack'а (
Система сборки позволила бы использовать модули.
Го тогда хотя бы просто вынесем это в отдельный файл, чтобы не было инлайн <script>

Sovianum replied:
Ок

Antonboom commented on html/leaders.html:
```diff
@@ -0,0 +1,63 @@
+<script src="/carousel-listener-other.js"></script>
+
+</body>
+</html>
```
В конце файла - пустая строка.
И мне кажется, что не во всех файлах все еще единый стиль отступов?

Meganster replied:
Переделаю

Antonboom commented on html/login.html:
```diff
@@ -0,0 +1,82 @@
+  <!-- End of container -->
+  
+  <!-- Core JavaScript-->
+  <script src="/polyfills.js"></script>
```
Давай указывать явно type="text/javascript"? Пусть для HTML5 это не обязательно.

Meganster replied:
Ок

Antonboom commented on html/signup.html:
```diff
@@ -0,0 +1,90 @@
+
+    <form class="form-horizontal form-fields">
+      <div class="form-group">
+        <label for="exampleInputName2">Name</label>
```
exampleInputName2 - ты серьезно?
Как насчет просто "name"

Meganster replied:
Переделаю

Antonboom commented on core/game_lib.js:
```diff
@@ -0,0 +1,454 @@
+
+
+
+function testCS() {
```
Вынесем все это туда, где это и должно быть - tests/game.js
И стандартного assertion должно хватить + сверху убрать можно будет константу.
А тесты на игру - круто.

Sovianum replied:
Вообще я собирался их совсем выпилить)

Antonboom replied:
Неееееет :D

Antonboom commented on core/game_lib.js:
```diff
@@ -0,0 +1,454 @@
+        this._radius = radius;
+    }
+
+    get radius() {
```
При желании такие вещи можно писать в одну строку - для экономии места.

Sovianum replied:
Предпочитаю оставлять так для единства стиля. Тем более, что об экономии места в любом случае позаботится минификатор.

Antonboom commented on core/game_lib.js:
```diff
@@ -0,0 +1,454 @@
+}
+
+
+class GameWorld {
```
Сюда можно пробрасывать _canvas - чтобы знать ее размеры и использовать в дальнейшем.

Sovianum replied:
В GameWorld не пробрасывал специально, чтобы была единственная точка связки с html-странцей - GameLoop. Лучше тогда в нем изменю логику инициации мира.

Antonboom commented on core/game_lib.js:
```diff
@@ -0,0 +1,454 @@
+}
+
+
+class GameLoop {
```
При желании, можно назвать просто Game и тогда будет game.start() :)

Sovianum replied:
Логично)

Antonboom reviewed (commented):
1. Во всех ли html-файлах единый стиль отступов
2. Хоть мы и используем модное class, хотелось бы, чтобы вы знали, как можно по-другому
https://learn.javascript.ru/classes

И как пользоваться тем, что имеем:
https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Classes

3. В игровую логику особо не вникал, изменения связанные с ней, можно оставить на следующую итерацию. Но "ок" в комментарии жду

4. В целом молодцы

Commit:
First playable example. Control platforms with arrows.

Commit:
merged conflicts

Commit:
Added core/lib exposure (temporarily, while haven't inserted math.js to game.js)

Commit:
v1

Commit:
1. Added events
2. Added size adaptation
3. Stopped working on the game cos it's uncomfortable to work without project assembler. Swotched to another task

Commit:
added webpack without grunt

Commit:
fixed assembly errors

Commit:
removed import

Commit:
Set up grunt

Commit:
Cosmetic fix

Commit:
Added cli

Commit:
Hacking

Commit:
Finally found

Commit:
1. Divided game logic into parts.
2. Another try to install grunt globally

Commit:
Trying to fool npm

Commit:
Rebuild please

Commit:
Another try

Commit:
Another try 1

Commit:
Another try 2

Commit:
Another try 3

Commit:
Another try 4

Commit:
Another try 5

Commit:
Another try 6

Commit:
Another try 7

Commit:
Another try 8

Commit:
Another try 9

Commit:
Another try 10

Commit:
Another try 11

Commit:
Another try 12

Commit:
Another try 13

Commit:
Another try 14

Commit:
Another try 15

Commit:
Another try 16

Commit:
Another try 17

Commit:
Another try 18

Commit:
Another try 19

Commit:
Enabled minification and added dynamic binding

Commit:
Merged design

Commit:
Cosmetic

Commit:
Set up event communication between game and server interface

Antonboom commented on Gruntfile.js:
```diff
@@ -0,0 +1,108 @@
+        jshint: {
+            all: ['core/*.js'],
+            options: {
+                "esversion": 6,
```
Там не было опции типа "путь к конфигу"? :)

Sovianum replied:
Опция, вроде, была, но по какой-то причине (уже не помню по какой) с ней у меня грант не работал. Могу использовать ее, но, вроде, раз мы собирается мигрировать на другой статический анализатор, перенастройка jshint неактуальна.

Antonboom replied:
Ок. Это я так, на заметку по уменьшению грантфайла)

Antonboom commented on Gruntfile.js:
```diff
@@ -0,0 +1,108 @@
+        'webpackStarted', 'webpack', 'watch'
+    ]);
+
+    grunt.registerTask('dev', ['webpackStarted', 'webpack', 'watch'])
```
Watch работает? Скорее всего его нужно сделать параллельным таском
См. grunt-concurrent
https://github.com/sindresorhus/grunt-concurrent

Sovianum replied:
Нет, watch не работает. Причем не работает он даже в том случае, когда я запускаю его последней командой, а потом стартую сервер вручную. Сегодня буду разбираться.

Antonboom replied:
Отпиши в диалог о результатах. Интересный момент.

Sovianum replied:
Без параллели заработал - неверно был задан паттерн поиска файлов.

Sovianum replied:
Сделал параллельным таском.

Antonboom commented on core/client_sever_dispatcher.js:
```diff
@@ -0,0 +1,28 @@
+    }
+}
+
+module.exports.ClientServerDispatcher = ClientServerDispatcher;
```
Здесь и далее.
Когда модуль представляет собой единственный класс (как впрочем и должно быть), то лучше писать так
```javascript
// client.js
let ClientServerDispatcher = class {...}
module.exports = ClientServerDipatcher;
```
Тогда можно будет писать
```javascript
// main.js
let ClientServer = require('./client');
let client = new ClientServer();
```
Аналог "export default".

Sovianum replied:
Спасибо. Исправил.

Antonboom commented on core/events.js:
```diff
@@ -0,0 +1,43 @@
+
+class PlatformMovedEvent {
+    static get eventName() {
+        return "PlatformMoved";
```
Лучше придерживаться именования констант "PLATFORM_MOVED_EVENT".
Плюс, в данном случае ты мог просто возвращать имя класса :)

Antonboom commented on core/events.js:
```diff
@@ -0,0 +1,43 @@
+}
+
+
+class DefeatEvent {
```
Слишком явно прослеживается одинаковая логика у нескольких классов.
Можно сделать общего родителя (в конструкторе параметр detail) и наследоваться от него. В event name отдавать имя унаследованного класса или добавить имя эвента в конструктор :)

```javascript
module.exports.PlatformMovedEvent = class PlatformMovedEvent extends BaseEvent {};
```

Sovianum replied:
1. Спасибо. В прошлый раз не дошло, поэтому сделал в лоб. Сейчас исправил.
2. Нет, если добавлять имя эвента в конструктор, то вся затея теряет смысл. Ведь в том-то все и дело, чтобы события однозначно идентифицировались их классом, а не содержащимся в них строковым полем (перепутать строку в процессе создания возможно, а класс - нет).

Antonboom commented on core/events.js:
```diff
@@ -0,0 +1,43 @@
+
+module.exports.PlatformMovedEvent = PlatformMovedEvent;
+module.exports.BallPositionCorrectionEvent = BallPositionCorrectionEvent;
+module.exports.DefeatEvent = DefeatEvent;
```
Здесь и далее - в конце файла пустая строка :)
Кажется, так и напрашивается ESLint

Sovianum replied:
Пустые строки добавил. ESLint постараюсь сделать в ближайшее время.

Antonboom replied:
Таска в трелле есть. При желании это можно назначить кому-нибудь. Не сложно все-таки.

Sovianum replied:
Поменял JShint на ESLint

Antonboom commented on core/game.js:
```diff
@@ -0,0 +1,232 @@
+
+const math = require('./_lib/math');
+const events = require('./events');
+const GameWorld = require('./game_world').GameWorld;
```
Писал об этом выше.
Ведь удобнее делать require() без обращения к атрибуту?)

Sovianum replied:
Исправил.

Antonboom commented on core/game.js:
```diff
@@ -0,0 +1,232 @@
+        this._setListeners();
+        this._initWorld();
+
+        let time = 1000 / this._frameRate;
```
Бывает, что такие вещи тоже выносят в константы вроде 1_SEC, чтобы повысить читаемость и не думать, в каких единицах измерения эта тысяча.
Так, для справки.

Sovianum replied:
Исправил.

Antonboom commented on core/game.js:
```diff
@@ -0,0 +1,232 @@
+    }
+
+    _handleKeyDown(event) {
+        if (event.keyCode == KEY_LEFT) {
```
Как вариант
```javascript
this._leftPressed = (event.keyCode == KEY_LEFT);
this._rightPressed = (event.keyCode == KEY_RIGHT);
```

Sovianum replied:
Отвечаю в посте ниже

Antonboom replied:
Ок, просто не забывай про форму приведенную выше.

Antonboom commented on core/game.js:
```diff
@@ -0,0 +1,232 @@
+    }
+
+    _handleKeyUp(event) {
+        if (event.keyCode == KEY_LEFT) {
```
А, не заметил, что ты разделил на up и down - в этом есть необходимость?

Sovianum replied:
У игрока есть две управляющие кнопки. Предположим следующую комбинацию их нажатия:
1. Нажать налево.
2. Нажать направо.
3. Отжать налево.
4. Отжать направо.
Мне кажется, что последовательность движения платформы должна быть следующая:
1) Двигаться налево.
2) Остановиться.
3) Двигаться направо.
4) Остановиться.
В этом случае состояние управляющей переменной нужно менять не только при нажатии, но и при отжатии клавиш.

Antonboom replied:
Да, так и думал. Ок.

Antonboom commented on core/game.js:
```diff
@@ -0,0 +1,232 @@
+    }
+
+    _updatePlatformVelocityDirection() {
+        if (!this._leftPressed && !this._rightPressed) {
```
Бывает, такие штуки превращают в более простое выражение, скорее, чтобы выиграть в скорости кода, чем в его читаемости.

Например, для этого случая искомая логическая функция выглядит так:
func(left, right) = position
func(0, 0) = [0, 0]		
func(0, 1) = [1, 0] 
func(1, 0) = [-1, 0] 
func(1, 1) = [0, 0] 

Получаем
```javascript
function getDirection(leftPressed, rightPressed) {
	if ( !(leftPressed ^ rightPressed) ) {     // XOR
		return [0, 0];
	}
	return [rightPressed - leftPressed, 0];
}
```
Если я нигде не напутал)) Но смысл понятен.

Sovianum replied:
Понимаю. Но мне кажется, пока лучше делать упор на читаемость. Ведь, как известно, микрооптимизации - корень всех зол)

Antonboom replied:
Ок, я просто фанат таких штук :)

Antonboom commented on core/game.js:
```diff
@@ -0,0 +1,232 @@
+
+    _handleDefeatEvent(event) {
+        // TODO сделать что-нибудь поинтереснее
+        let sectorId = event.detail;
```
Например, отправить event фронту, чтобы он вывел сообщение о том, что игра завершена и пр..
Ну и в дальнейшем, евент на бэкенд.

Sovianum replied:
На бэкенд - это вряд ли. Я думал, что события о проигрыше и коррекции приходят только от бэка фронту. Фронт же шлет только события о перемещении платформ.
Да, нужно будет еще добавить событие для перемещения вражеских ракеток.

Antonboom replied:
Да, это я приболел немного. События, конечно, от бэкенда фронту :)

Antonboom commented on core/game.js:
```diff
@@ -0,0 +1,232 @@
+        } else {
+            alert("You win");
+        }
+        this._world.userSectors[event.detail].setLoser();
```
Выше определен sectorId = event.detail.

Sovianum replied:
Спасибо. Исправил.

Antonboom commented on core/game_components.js:
```diff
@@ -0,0 +1,256 @@
+        ];
+
+        let result = true;
+        result = result && this.leftBorder <= checkPoint[0];
```
Сложна.
```javascript
result = (this.leftBorder <= checkPoint[0] <= this.rightBorder) && 
         (this.lowerBorder <= checkPoint[1] <= this.upperBorder);
```
Вроде бы JS в такое умел, если нет, то я плак-плак 😢

Sovianum replied:
Похоже, что нет.
Следующий код у меня выводит true:
`(x => { return 0 < x < 100; })(-10) // true`

Antonboom replied:
Тогда я плак-плак.
Может, хотя бы, все-таки в одно выражение на разных строках?
Сколько переприсваиваний приходится делать интерпретатору и каждый раз создается новый объект :(

Sovianum replied:
Поправил.

Antonboom commented on core/game.js:
```diff
@@ -0,0 +1,232 @@
+    }
+
+    get _playerItemsId() {
+        return Math.floor(this._playerNum / 2);
```
Может написать, как это работает? :)
Особо не вчитывался, но по названию не могу понять, что такое id итемсов игрока и почему ты получаешь его делением на два от номера игрока?)

Sovianum replied:
Дело в том, что у меня нумерация секторов и платформ идет сверху по часовой стрелке. А пользователь находится на нижнем сегменте. Так я получаю индекс, по которому можно получить сектор и платформу данного игрока. Видимо название метода неудачное. Нет мыслей, на что можно переименовать?

Antonboom replied:
Столку сбило _playerNum - не понятно, что это множественное число, а в конструкторе проглядел. _playersNum тогда уж. 
И ты получаешь не ID, а как бы индекс в массиве платформ же?
get _currentPlayerSectorID
get _currentPlayerSectorIndex
get _playerSectorIdx
get _currentPlayerSector  <-- збс

В общем _playerNum переименуй, с геттером решай сам.
get _playerItemsIndex

Сложна :)

Sovianum replied:
Исправил.

Antonboom reviewed (commented):
1. Где мне ожидать код от Юры?
2. Куда пропали тесты игрового мира? :)
3. Расскажи, где так быстро учат круто прогать на канвасе, я тоже хочу.

:1st_place_medal:

Commit:
Small cosmetic fixes

Antonboom commented on package.json:
```diff
@@ -15,7 +16,33 @@
   "author": "",
   "homepage": "https://polyhedron.herokuapp.com/",
   "devDependencies": {
```
Лучше сделай так:
https://devcenter.heroku.com/articles/nodejs-support#devdependencies

И все-таки разграничь дев и обычные зависимости :)

Sovianum replied:
Сделал

Antonboom commented on package.json:
```diff
@@ -15,7 +16,33 @@
   "author": "",
   "homepage": "https://polyhedron.herokuapp.com/",
   "devDependencies": {
```
А лучше сделай так:
https://devcenter.heroku.com/articles/nodejs-support#devdependencies

И все-таки разграничь дев и обычные зависимости :)

Commit:
1. Fixed game import bugs;
2. Set up concurrent grunt watch;
3. Set dev mode on heroku;
4. Cut out redundant production dependencies;

Commit:
Cosmetic

Commit:
Added dev webpack dependency

Commit:
Fixed grunt postinstall

Commit:
Changed jshint to eslint

Commit:
Enabled minification

Antonboom reviewed (approved):
Красота.
